### ISSUE TITLE
Add2App [Option A&B] with plugins failed to build because plugins was built without full bitcode

### ISSUE THREAD

**Issue by nailgilaziev** [flutter/flutter#48092]
This is a next step of an issue described here #47930 and all preparation steps is the same.
But now before building `.xcframework` just add to `my_flutter` module some plugin dependencies. 
I added most popular _shared_preferences_ plugin to `pubspec.yaml` and added example code to `main.dart` file.
Next
```
flutter build ios-framework --output=../host_ios/embed/ --xcframework --no-universal
```
produce 
```
App.framework
Flutter.xcframework
FlutterPluginRegistrant.xcframework
shared_preferences.xcframework
```
Applying them to xcode pbxproj -> build -> error

> ld: bitcode bundle could not be generated because '/Users/n/Library/Developer/Xcode/DerivedData/host_ios-bnbwehcmdjzqgydxspnsnfifvdyh/Build/Intermediates.noindex/ArchiveIntermediates/host_ios/BuildProductsPath/Release-iphoneos/shared_preferences.framework/shared_preferences' was built without full bitcode. All frameworks and dylibs for bitcode must be generated from Xcode Archive or Install build file '/Users/n/Library/Developer/Xcode/DerivedData/host_ios-bnbwehcmdjzqgydxspnsnfifvdyh/Build/Intermediates.noindex/ArchiveIntermediates/host_ios/BuildProductsPath/Release-iphoneos/shared_preferences.framework/shared_preferences' for architecture arm64

I checked this situation with another plugin - _path_provider_ - result the same.

If I try to disable *bitcode* I get another error:
![headerr](https://user-images.githubusercontent.com/2749021/71697211-54583f00-2dc8-11ea-87c9-829a90a2d344.png)
with this logs:
[IDEDistribution.critical.log](https://github.com/flutter/flutter/files/4017571/IDEDistribution.critical.log)
[IDEDistribution.standard.log](https://github.com/flutter/flutter/files/4017572/IDEDistribution.standard.log)
[IDEDistribution.verbose.log](https://github.com/flutter/flutter/files/4017573/IDEDistribution.verbose.log)

**Comment by nailgilaziev** [comment]
This issue was on 
`Flutter (Channel master, v1.13.6-pre.48, on Mac OS X 10.15.2 19C57, locale en-RU)`
upgraded ~ 3 hour ago, but after 1 hour new version released, so...

additionally I upgraded to currently latest:
```
Flutter 1.13.7-pre.8 • channel master • https://github.com/flutter/flutter.git
Framework • revision dc99d61e5d (2 hours ago) • 2020-01-02 13:03:01 -0800
Engine • revision fa05d367ad
Tools • Dart 2.8.0 (build 2.8.0-dev.0.0 b52d264f5b)
```
and repeat test:
Result the same - bitcode error
PS now building frameworks generate _App.xcframework_, not the _App.framework_ as before.

**Comment by nailgilaziev** [comment]
You doesn’t apply _severe_ label. 
Any workaround exist for this problem? If yes - I will apply it for publishing my app to the TestFlight.

**Comment by jmagman** [comment]
I need to try this, I'm not sure what's going on and don't have a workaround at the moment.

**Comment by Magsumov2** [comment]
Hello! @jmagman , please tell me - When will it be fixed? I meet the same issue and I can not release application because of this error! thanks

**Comment by nailgilaziev** [comment]
@jmagman , I tried to add flutter module to app via Option A - cocoapods script, recommended way
But when archive this error occurs:
> bitcode bundle could not be generated because '/Users/n/pubios/my_flutter/.ios/Flutter/engine/Flutter.framework/Flutter' was built without full bitcode. All frameworks and dylibs for bitcode must be generated from Xcode Archive or Install build file '/Users/n/pubios/my_flutter/.ios/Flutter/engine/Flutter.framework/Flutter' for architecture armv7

**Comment by nailgilaziev** [comment]
to make sure I'm doing everything right and to provide more information for you I do everything from scratch with latest flutter version:
```
[✓] Flutter (Channel master, v1.13.8-pre.58, on Mac OS X 10.15.2 19C57, locale en-RU)
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
[✓] Android Studio (version 3.5)
```

I create new flutter module 
`flutter create --template module my_flutter`
create new xcode project and then configure pod file and reopen _.xcworkspace_

with simplest flutter module everything compiles and successfully uploads to AppStore. 
But when I add dependency to flutter module and try to build and upload again I meet the error on archive:
> bitcode bundle could not be generated because ... was built without full bitcode....

![2err](https://user-images.githubusercontent.com/2749021/72075078-15644500-3304-11ea-87b2-9798b3a93569.png)

**Comment by jmagman** [comment]
Need integration tests that not just build the host app, but archive and validate it, too.

1. Found an unexpected Mach-O header code: 0x72613c21
```
2020-01-14 02:56:03 +0000  Failed to generate distribution items with error: Error Domain=DVTMachOErrorDomain Code=0 "Found an unexpected Mach-O header code: 0x72613c21" UserInfo={NSLocalizedDescription=Found an unexpected Mach-O header code: 0x72613c21, NSLocalizedRecoverySuggestion=}
```
Xcode project settings > Build Settings > Embed Frameworks and remove FlutterPluginRegistratant.framework (or xcframework).  Keep it in Link Binary With Libraries.

This is a static framework, not a dynamic one...
I will update the docs.  Hopefully this unblocks you.

2. Plugin framework was built without full bitcode. 
```
ld: bitcode bundle could not be generated because '/Users/n/Library/Developer/Xcode/DerivedData/host_ios-bnbwehcmdjzqgydxspnsnfifvdyh/Build/Intermediates.noindex/ArchiveIntermediates/host_ios/BuildProductsPath/Release-iphoneos/shared_preferences.framework/shared_preferences' was built without full bitcode. All frameworks and dylibs for bitcode must be generated from Xcode Archive or Install build file '/Users/n/Library/Developer/Xcode/DerivedData/host_ios-bnbwehcmdjzqgydxspnsnfifvdyh/Build/Intermediates.noindex/ArchiveIntermediates/host_ios/BuildProductsPath/Release-iphoneos/shared_preferences.framework/shared_preferences' for architecture arm64
```
I believe this one is because the plugins are being built instead of archived, and so aren't being created with bitcode.  I'm about to put out a pull request to fix this one.

3. Flutter was built without full bitcode
```
bitcode bundle could not be generated because '/Users/n/pubios/my_flutter/.ios/Flutter/engine/Flutter.framework/Flutter' was built without full bitcode. All frameworks and dylibs for bitcode must be generated from Xcode Archive or Install build file '/Users/n/pubios/my_flutter/.ios/Flutter/engine/Flutter.framework/Flutter' for architecture armv7
```
I believe this could happen if `flutter build --debug` was last run in the module and there's a debug version of Flutter at the `vendored_framework` path.  The release version *should* be re-copied when the `[CP-User] Run Flutter Build Script` build phase runs in the host app, though.  I need to look into this one more.
In the meantime, try running `flutter build --release` in your Flutter module, then re-try `pod install` from your host app.

**Comment by nailgilaziev** [comment]
@jmagman, Does you(or team) need any tests or clarifications from me?
This issue is blocking and no other ways to workaround it. 
Adding flutter module with some pub dependencies to host app currently not working at all. 
Does not work in either the stable or master channel. neither option **A** nor **B**
Please help.

**Comment by jmagman** [comment]
@nailgilaziev Did you try my suggestions above?  

For option A (CocoaPods) try running `flutter build ios --release` in the Flutter module.

For option B (generating frameworks) disable bitcode  in your host app, and remove FlutterPluginRegistratant.framework (or xcframework) from the Embed Frameworks build phase.  It's a static framework and so should be linked on and not embedded.  Depending on what plugins you have, you may also need to do the same for them.  Run:
```
$ grep -riI "static_framework" .ios/Pods
```
to see which podspecs (CocoaPods metadata per plugin) should also be removed.

**Comment by nailgilaziev** [comment]
@jmagman , sorry, I didn't saw you message when writing "help" :) my bad. I will try it now and get feedback today

**Comment by nailgilaziev** [comment]
yeah, option B with excluded FlutterPluginRegistratant.framework (or xcframework) from Embed Frameworks phase is working! 
Thanks! I already released long-awaited release!

**Comment by jmagman** [comment]
Added the warning here:
https://flutter.dev/docs/development/add-to-app/ios/project-setup#embed-the-frameworks

**Comment by jmagman** [comment]
> @jmagman , I tried to add flutter module to app via Option A - cocoapods script, recommended way
> But when archive this error occurs:
> 
> > bitcode bundle could not be generated because '/Users/n/pubios/my_flutter/.ios/Flutter/engine/Flutter.framework/Flutter' was built without full bitcode. All frameworks and dylibs for bitcode must be generated from Xcode Archive or Install build file '/Users/n/pubios/my_flutter/.ios/Flutter/engine/Flutter.framework/Flutter' for architecture armv7

@nailgilaziev  https://github.com/flutter/flutter/pull/49102 should fix this one on the [master channel](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels).  If you happen to get a chance to validate that it's fixed, that would be greatly appreciated.

In any case, thank you for this report!  You helped us fix it quickly.  😄

**Comment by nailgilaziev** [comment]
@jmagman , no, Option A is not working currently. Maybe I do something wrong.
I do all steps from scratch for sure.
upgraded to
```
[✓] Flutter (Channel master, v1.14.4-pre.17, on Mac OS X 10.15.2 19C57, locale en-RU)
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
[✓] Android Studio (version 3.5)
```


```
cd pubios/
flutter create --template module fmodule
```
* opened project in androidStudio, add shared_pref dependency, did `flutter pub get` and  add example code for using it
* created new xcode proj `podhost` in _pubios_ dir
```
cd podhost/
pod init 
```
Podfile content is next:
```
# Uncomment the next line to define a global platform for your project
platform :ios, '9.0'

flutter_application_path = '../fmodule'
 load File.join(flutter_application_path, '.ios', 'Flutter', 'podhelper.rb')

target 'podhost' do
  # Comment the next line if you don't want to use dynamic frameworks
  use_frameworks!

  # Pods for host_cocoa
  install_all_flutter_pods(flutter_application_path)
end
```
```
pod install 
```
produce this
```
Analyzing dependencies
Downloading dependencies
Installing Flutter (1.0.0)
Installing FlutterPluginRegistrant (0.0.1)
Installing fmodule (0.0.1)
Installing shared_preferences (0.0.1)
Installing shared_preferences_macos (0.0.1)
Installing shared_preferences_web (0.0.1)
Generating Pods project
Integrating client project

[!] Please close any current Xcode sessions and use `podhost.xcworkspace` for this project from now on.
Pod installation complete! There are 6 dependencies from the Podfile and 6 total pods installed.
```

```
pro:podhost n$ open podhost.xcworkspace/
```
Open xcode and try to archive
![errshared](https://user-images.githubusercontent.com/2749021/72886535-083f5100-3d1b-11ea-9cb8-ea0d545eafc7.png)

When creating xcode project default settings in project target for _Build Options_->_Enable Bitcode_ is *Yes*(true)
Add2App documentation says anything about this. Do we need set it to *No*(false)? 

Pods targets had a similar value for _Build Options_->_Enable Bitcode_ is *Yes*(true)

Even if I set _Enable Bitcode_ to  *No*(false) for project target - the error remains

**Comment by jmagman** [comment]
I see the problem, and it's similar to https://github.com/flutter/flutter/issues/37850.

When I run `flutter build ios --release` in the module, the host app can then archive successfully.  When I run `flutter build ios --debug` in the module, we see the reported error on archive.

The "Run Flutter Build Script" host app script build phase should be swapping the right version of the engine based on the host app build configuration (Debug, Release), but:
1.  The plugin target (shared_preferences) is being built before the host app, so before that build phase can run.
2. The script build phase has Input files from the module, so won't run every time.
<img width="672" alt="input" src="https://user-images.githubusercontent.com/682784/72926712-be01a400-3d09-11ea-9d67-652197538069.png">

**Comment by isacjunior** [comment]
@jmagman I believe that is important to add in website(docs) to recommend run `flutter build` before `pod install` for these reasons above and for another reason that I'll explain below:

When I create a new flutter module, flutter script will create a `.ios` folder and run `pod install`, the script inside `podhelper.rb` will create an `App.framework` used in Pods only with arch x86_64.

**To reproduce:**

```bash
flutter create --template=module example
cd ios_app
pod install
```
Because `app_framework_dylib` doesn't exist will create `App.framework`.
```bash
`echo "static const int Moo = 88;" | xcrun clang -x c -dynamiclib -o "#{app_framework_dylib}" -`
```
When I use `File` to show more details, this is output:
```
App: Mach-O 64-bit dynamically linked shared library x86_64
```
For resolving this, run build with the mode that you need, for example:
```bash
flutter build ios --release --no-codesign
```
When I use `File` now, these are output:
```
App: Mach-O universal binary with 2 architectures: [arm_v7:Mach-O dynamically linked shared library arm_v7] [arm64]
App (for architecture armv7):	Mach-O dynamically linked shared library arm_v7
App (for architecture arm64):	Mach-O 64-bit dynamically linked shared library arm64
```

**Question:**
For a particular reason, I need architecture x86_64 when I run in `Release` mode, Is possible?

**Doctor:**
```
[✓] Flutter (Channel unknown, v1.14.4, on Mac OS X 10.14.6 18G2022, locale en-BR)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
[✓] Android Studio (version 3.5)
[✓] VS Code (version 1.41.1)
[✓] Connected device (1 available)
```

**Comment by jmagman** [comment]
> For a particular reason, I need architecture x86_64 when I create the framework with Release mode, it possible? Some workaround?

Flutter is not supported in the simulator (x86_64) in release mode. See https://github.com/flutter/flutter/issues/11754.

**Comment by l3ender** [comment]
Hello, just wanted to drop a quick note thanking you for the detail in this issue and the related PR. We are not flutter users, but came across similar iOS build errors in our app that links against frameworks from a different project. The frameworks are delivered after being built (not archived), so the fix to build with `BITCODE_GENERATION_MODE=bitcode` was very helpful for us. Thank you!

**Comment by Sauraus** [comment]
> Hello, just wanted to drop a quick note thanking you for the detail in this issue and the related PR. We are not flutter users, but came across similar iOS build errors in our app that links against frameworks from a different project. The frameworks are delivered after being built (not archived), so the fix to build with `BITCODE_GENERATION_MODE=bitcode` was very helpful for us. Thank you!

Where and how do you set this?

**Comment by RDyego** [comment]
> > Hello, just wanted to drop a quick note thanking you for the detail in this issue and the related PR. We are not flutter users, but came across similar iOS build errors in our app that links against frameworks from a different project. The frameworks are delivered after being built (not archived), so the fix to build with `BITCODE_GENERATION_MODE=bitcode` was very helpful for us. Thank you!
> 
> Where and how do you set this?
Look on xcode:
![image](https://user-images.githubusercontent.com/6733336/79765013-784e7a80-82fc-11ea-96e0-267075122b5b.png)

**Comment by Sumit-Ghosh** [comment]
Even I am facing this bitcode issue with **webview_flutter** plugin

**Comment by bawantha** [comment]
> .

same

**Comment by bawantha** [comment]
> 
> 
> Even I am facing this bitcode issue with **webview_flutter** plugin.
> I am using option A that is added to the app via Cocoapods.
> 
> macOS: 10.15.3 (Catalina)
> Xcode version: 11.0 (11A420a)
> Flutter version: Channel stable, 1.20.2

same

**Comment by jmagman** [comment]
https://github.com/flutter/flutter/pull/73072 is available on 1.26.0-12.0.pre, which is currently available on the [master, dev, and beta channels](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#how-to-change-channels).

**Comment by uerceg** [comment]
Hi @jmagman 

Sorry for ping on closed issue, but seems like I am still facing this even with latest stable build:

<details>
    <summary>flutter doctor -v output</summary>

```
[✓] Flutter (Channel stable, 2.0.2, on macOS 11.2.3 20D91 darwin-x64, locale en-GB)
    • Flutter version 2.0.2 at /Users/uerceg/Development/flutter
    • Framework revision 8962f6dc68 (7 days ago), 2021-03-11 13:22:20 -0800
    • Engine revision 5d8bf811b3
    • Dart version 2.12.1

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/uerceg/Development/android/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/uerceg/Development/android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] IntelliJ IDEA Community Edition (version 2020.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 53.1.5
    • Dart plugin version 203.6912

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (3 available)
    • Pixel 4a 5G (mobile)      • 0B061JECB12057            • android-arm64  • Android 11 (API 30)
    • Uglješa’s iPhone (mobile) • 00008030-0011104A228B802E • ios            • iOS 14.5
    • Chrome (web)              • chrome                    • web-javascript • Google Chrome 89.0.4389.90
    ! Error: Living Room is not connected. Xcode will continue when Living Room is connected. (code -13)

• No issues found!
```
</details>

We're developing the plugin which is added to the app where this is reproducible. You can find the app project in [here](https://drive.google.com/file/d/1nFsMU6wVzi__WXW0eaGmK68Z85ILhy-F/view?usp=sharing) in case you want to reproduce this.

As described in [our repo issue](https://github.com/adjust/flutter_sdk/issues/42), using **Release** build configuration does the trick as workaround to be able to run the app on iOS device. Is this still supposed to be like that?

Cheers

**Comment by jmagman** [comment]
@uerceg https://github.com/adjust/flutter_sdk/issues/42 sounds like it's a normal Flutter app trying to consume your plugin, not necessarily [add-to-app](https://flutter.dev/docs/development/add-to-app) related.  I just ran your [sample app](https://drive.google.com/file/d/1nFsMU6wVzi__WXW0eaGmK68Z85ILhy-F/view) on a phone against Flutter stable and it built and launched fine in debug mode, I don't see any bitcode related warnings.

It's running:
```
2021-03-18 18:38:55.220440-0700 Runner[1752:1764709] 	[Adjust]w: SANDBOX: Adjust is running in Sandbox mode. Use this setting for testing. Don't forget to set the environment to `production` before publishing
...
```
It doesn't even have bitcode enabled (which aligns with the default in a `flutter create`d app):
<img width="364" alt="Screen Shot 2021-03-18 at 6 42 06 PM" src="https://user-images.githubusercontent.com/682784/111719701-a6dd1a00-8819-11eb-863f-2694e4e9df3e.png">

I recommend you file a new issue since it doesn't look related to add-to-app.

**Comment by uerceg** [comment]
Huh! Really interesting. Thank you for lightning speed check on this one. 🍺 I opened new [issue](https://github.com/flutter/flutter/issues/78589) for this.
